tests/mocks/core/services/request/RequestMock.php

```diff
@@ -0,0 +1,211 @@
+    public $request_type;
+
+    /**
+     * IP address for request
+     *
+     * @var string $ip_address
+     */
+    public $ip_address;
+
+    /**
+     * @var string $user_agent
+     */
+    public $user_agent;
+
+    /**
+     * true if current user appears to be some kind of bot
+     *
+     * @var bool $is_bot
+     */
+    public $is_bot;
+
+
+    public function __construct(array $get, array $post, array $cookie, array $server, $ip_address = '0.0.0.0')
+    {
+        $this->get = $get;
+        $this->post = $post;
+        $this->cookie = $cookie;
+        $this->server = $server;
+        $this->request = array_merge($this->get, $this->post);
+        $this->ip_address = $this->visitorIp($ip_address);
+        parent::__construct($get, $post, $cookie, $server);
+    }
+
+
+    /**
+     * @param string $key
+     * @param string $value
+     * @param bool   $override_ee
+     * @return    void
+     */
+    public function setRequestParam($key, $value, $override_ee = false)
+    {
+        $this->request[ $key ] = $value;
+    }
+
+
+    /**
+     * returns the value for a request param if the given key exists
+     *
+     * @param string $key
+     * @param null   $default
+     * @return mixed
+     */
+    public function getRequestParam($key, $default = null)
+    {
+        return isset($this->request[ $key ]) ? $this->request[ $key ] : $default;
+    }
+
+
+    /**
+     * check if param exists
+     *
+     * @param string $key
+     * @return bool
+     */
+    public function requestParamIsSet($key)
+    {
+        return isset($this->request[ $key ]);
+    }
+
+
+    /**
+     * check if a request parameter exists whose key that matches the supplied wildcard pattern
+     * and return the value for the first match found
+     * wildcards can be either of the following:
+     *      ? to represent a single character of any type
+     *      * to represent one or more characters of any type
+     *
+     * @param string     $pattern
+     * @param null|mixed $default
+     * @return false|int
+     */
+    public function getMatch($pattern, $default = null)
+    {
+        // replace wildcard chars with regex chars
+        $pattern = str_replace(
+            array("\*", "\?"),
+            array('.*', '.'),
+            preg_quote($pattern, '/')
+        );
+        foreach ($this->request as $key => $request_param) {
+            if (preg_match('/^' . $pattern . '$/is', $key)) {
+                return $request_param;
+            }
+        }
+        return null;
+    }
+
+
+    /**
+     * check if a request parameter exists whose key matches the supplied wildcard pattern
+     * wildcards can be either of the following:
+     *      ? to represent a single character of any type
+     *      * to represent one or more characters of any type
+     * returns true if a match is found or false if not
+     *
+     * @param string $pattern
+     * @return false|int
+     */
+    public function matches($pattern)
```

Comment:
Interestingly enough, in looking at this code it seems to suggest production code could actually be modified to call `getMatch` instead of `requestParameterDrillDown`.

Reply:
well you're not wrong...

but look at the actual code involved:

```php
    public function getMatch($pattern, $default = null)
    {
        return $this->requestParameterDrillDown($pattern, $default, 'match');
    }

    public function matches($pattern)
    {
        return $this->requestParameterDrillDown($pattern, null, 'match') !== null;
    }
```

if you changed `matches()` to call `getMatch()` instead, you would **still** be calling `requestParameterDrillDown()`, but just indirectly... 
so what's the point? You gain nothing from doing that.